### PR TITLE
More Flickr testing, this time on the import side

### PR DESCRIPTION
**Pull request by olsona**
Started on testing the FlickrPhotoService from the import side.  Needed to use an actual JobDataCache for readability.

### CODE REVIEW

File: portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java [hunk 1]
```diff
@@ -93,6 +106,40 @@ public void toCommonPhoto() {
...
+
+    // Verify the mocks were called correctly
+    verify(imageStreamProvider).get(FETCHABLE_URL);
+    verify(uploader)
+        .upload(Matchers.any(BufferedInputStream.class), Matchers.any(UploadMetaData.class));
+    verify(photosetsInterface).create(flickrAlbumTitle, ALBUM_DESCRIPTION, FLICKR_PHOTO_ID);
+
+    // Check jobDataCache contents
+    String expectedAlbumKey = FlickrPhotoService.CACHE_ALBUM_METADATA_PREFIX + ALBUM_ID;
+    assertThat(jobDataCache.hasKey(expectedAlbumKey));
```
Comment (rtannenbaum) [inline portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java]: do you need .isTrue()?

File: portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java [hunk 2]
```diff
@@ -93,6 +106,40 @@ public void toCommonPhoto() {
...
+    when(photosetsInterface.create(flickrAlbumTitle, ALBUM_DESCRIPTION, FLICKR_PHOTO_ID))
+        .thenReturn(photoSet);
+
+    // Run test
+    photoService.importItem(wrapper);
+
+    // Verify the mocks were called correctly
+    verify(imageStreamProvider).get(FETCHABLE_URL);
+    verify(uploader)
+        .upload(Matchers.any(BufferedInputStream.class), Matchers.any(UploadMetaData.class));
```
Comment (rtannenbaum) [inline portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java]: do you want to verify it used your inputStream specifically? (same thing in the when() call)

File: portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java [hunk 3]
```diff
@@ -93,6 +106,40 @@ public void toCommonPhoto() {
...
+
+    String flickrAlbumTitle = FlickrPhotoService.FLICKR_ALBUM_PREFIX + ALBUM_NAME;
+    Photoset photoSet = initializePhotoset(FLICKR_ALBUM_ID, flickrAlbumTitle, ALBUM_DESCRIPTION);
+    when(photosetsInterface.create(flickrAlbumTitle, ALBUM_DESCRIPTION, FLICKR_PHOTO_ID))
+        .thenReturn(photoSet);
+
+    // Run test
+    photoService.importItem(wrapper);
+
+    // Verify the mocks were called correctly
```
Comment (rtannenbaum) [inline portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java]: nit: might rephrase this as what we specifically want to test was called, e.g.

"Verify the photo was uploaded" etc

File: portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java (hunk 2 above)
Comment (rtannenbaum) [inline portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java]: Not sure if you're aware, another cool thing you can do with mockito is ArgumentCaptors. In this case I think it'd be useful to capture the uploadMetadata that is passed to uploader, and verify the title is what you expect.

File: portability-core/src/main/java/org/dataportabilityproject/cloud/local/InMemoryJobDataCache.java [hunk 4]
```diff
@@ -24,7 +24,7 @@
 import java.util.HashMap;
 import org.dataportabilityproject.cloud.interfaces.JobDataCache;
 
-final class JobDataCacheImpl implements JobDataCache {
+final public class InMemoryJobDataCache implements JobDataCache {
```
Comment (rtannenbaum) [inline portability-core/src/main/java/org/dataportabilityproject/cloud/local/InMemoryJobDataCache.java]: nit: order should be "public final"

File: portability-core/src/main/java/org/dataportabilityproject/shared/ImageStreamProvider.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package org.dataportabilityproject.shared;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+public class ImageStreamProvider {
```
Comment (rtannenbaum) [inline portability-core/src/main/java/org/dataportabilityproject/shared/ImageStreamProvider.java]: nit: can you add documentation for the class, as well as the license header?

File: portability-core/src/main/java/org/dataportabilityproject/dataModels/photos/PhotoModel.java [hunk 6]
```diff
@@ -54,4 +56,31 @@ public String getMediaType() {
     public String getAlbumId() {
         return albumId;
     }
+
```
[the remark this answers is not in the document]
Comment (olsona) [reply portability-core/src/main/java/org/dataportabilityproject/dataModels/photos/PhotoModel.java]: Separate PR it is.  Any idea why these weren't AutoValues in the first place?

File: portability-core/src/main/java/org/dataportabilityproject/shared/ImageStreamProvider.java (hunk 5 above)
Comment (olsona) [reply portability-core/src/main/java/org/dataportabilityproject/shared/ImageStreamProvider.java]: Adding light documentation.  I imagine more documentation will be added as the class grows.

File: portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java (hunk 2 above)
Comment (olsona) [reply portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java]: Unfortunately, BufferedInputStream.equals and UploadMetaData.equals both compare references.

Comment (olsona) [reply portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java]: Sweet!

File: portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java (hunk 1 above)
Comment (olsona) [reply portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java]: Yes, yes I do.

**Comment by olsona** [comment]
Ok, addressed the comments, and cleaned up the code some.

File: portability-core/src/main/java/org/dataportabilityproject/dataModels/photos/PhotoModel.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (rtannenbaum) [reply portability-core/src/main/java/org/dataportabilityproject/dataModels/photos/PhotoModel.java]: Don't know, but I don't see a reason it shouldn't be.

File: portability-core/src/main/java/org/dataportabilityproject/shared/ImageStreamProvider.java (hunk 5 above)
Comment (rtannenbaum) [reply portability-core/src/main/java/org/dataportabilityproject/shared/ImageStreamProvider.java]: can you make it a bit more descriptive? like what the stream is used for etc.

File: portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java (hunk 2 above)
Comment (rtannenbaum) [reply portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java]: I think we want that, actually -- we can easily verify that it got the right inputStream (i.e. your mock)

Comment (rtannenbaum) [reply portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java]: Nice! So for the other fields, this may be debatable but I don't think checking the values of those flags (isPublic etc) adds much value because they are hardcoded to false, whereas with the title we are checking some (albeit simple) logic from the class, i.e. the string concatenation and that the correct photo is passed.

Checking the other flag values might be like a change-detector test a la TOTT

File: portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java (hunk 1 above)
Comment (rtannenbaum) [reply portability-core/src/test/java/org/dataportabilityproject/serviceProviders/flickr/FlickrPhotoServiceTest.java]: thanks. Also the argumentCaptor checks you added use assertEquals... can you make those use assertThat's to be consistent w/ the others?

File: portability-core/src/main/java/org/dataportabilityproject/shared/ImageStreamProvider.java (hunk 5 above)
Comment (rtannenbaum) [reply portability-core/src/main/java/org/dataportabilityproject/shared/ImageStreamProvider.java]: Sorry to nitpick but I think it is still unclear to a reader what an image stream is exactly.. How about something like this:

Gets an input stream to an image, given its URL. Used by {@link FlickrPhotoService} to upload the image.
